pkg/backend/httpstate/client/marshal.go

```diff
@@ -16,20 +16,240 @@ package client
 
 import (
 	"bytes"
-	"encoding/json"
+	"compress/gzip"
+	"fmt"
+	"io"
+
+	"github.com/json-iterator/go"
 
 	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
 )
 
-// Unlike json.Marshal preserves possible indentation in req.Deployment.
-func marshalVerbatimCheckpointRequest(req apitype.PatchUpdateVerbatimCheckpointRequest) (json.RawMessage, error) {
-	sentinel := []byte(`"*"`)
-	cp := req
-	cp.UntypedDeployment = sentinel
-	pattern, err := json.Marshal(cp)
+// Implements io.WriterTo that writes canonical JSON in the apitype.UntypedDeployment format.
+//
+// Optimized for large checkpoints. Streams the data instead of buffering everything upfront.
+//
+// Injects newlines to allow efficient textual diffs over the JSON.
+func MarshalUntypedDeployment(deployment *apitype.DeploymentV3) io.WriterTo {
+	return &marshalUntypedDeployment{deployment}
+}
+
+type marshalUntypedDeployment struct {
+	deployment *apitype.DeploymentV3
+}
+
+var _ io.WriterTo = &marshalUntypedDeployment{}
+
+func (c *marshalUntypedDeployment) WriteTo(w io.Writer) (int64, error) {
+	cw := &countingWriter{w: w}
+	cfg := jsoniter.ConfigCompatibleWithStandardLibrary
+	stream := cfg.BorrowStream(cw)
+	defer cfg.ReturnStream(stream)
+	err := c.writeToStream(stream)
+	return cw.written, err
+}
+
+func (c *marshalUntypedDeployment) writeToStream(stream *jsoniter.Stream) error {
+	stream.WriteObjectStart()
+	stream.WriteObjectField("version")
+	stream.WriteInt(3)
+	stream.WriteMore()
+	stream.WriteObjectField("deployment")
+	c.writeDeploymentV3(stream)
+	stream.WriteObjectEnd()
+	return stream.Flush()
+}
+
+func (c *marshalUntypedDeployment) writeDeploymentV3(stream *jsoniter.Stream) (err error) {
+	deployment := c.deployment
+	stream.WriteObjectStart()
+	stream.WriteObjectField("manifest")
+	stream.WriteVal(deployment.Manifest)
+	if deployment.SecretsProviders != nil {
+		stream.WriteMore()
+		stream.WriteObjectField("secrets_providers")
+		stream.WriteVal(deployment.SecretsProviders)
+	}
+	if err = stream.Flush(); err != nil {
+		return err
+	}
+	if len(deployment.Resources) > 0 {
+		stream.WriteMore()
+		stream.WriteObjectField("resources")
+		stream.WriteRaw("[\n")
+		for i, r := range deployment.Resources {
+			if i > 0 {
+				stream.WriteRaw(",\n")
```

Comment:
We need to put newlines somewhere for text diff to work efficiently, this is one way to put some newlines, but it's different from the status quo. It's been disappointing to find issues with MarshalIndent.

`json.MarhalIndent(x, "", "")` used previously uses two-passes with in-mem buffers, expensive, but produces the desired result of many newlines with no extra space.

`json.NewEncoder(w); enc.SetIndent("", ""); enc.Encode(msg)` is inconsistent with `json.MarhalIndent` as it decides not to include the newlines at all.

Using `enc.SetIndent("", " ")` introduces unwanted whitespace besides the newlines.

`jsoniter` indent options never match the behavior of `json.MarhalIndent(x, "", "")`. It's more like `enc.SetIndent`. It either writes newlines and whitespace, or no newlines at all.